src/OrleansRuntime/GrainTypeManager/GrainTypeData.cs

```diff
@@ -151,5 +154,11 @@ internal static PlacementStrategy GetPlacementStrategy(Type grainClass, Placemen
 
             return predicate;
         }
+
+        private ParameterInfo[] GetConstructorParameters()
+        {
+            var constructor = this.Type.GetConstructors(BindingFlags.Public | BindingFlags.Instance).FirstOrDefault();
+            return constructor?.GetParameters() ?? EmptyParameters;
+        }
```

Comment:
My reasoning for moving it out isn't firm, I just noted that `GrainTypeData` is `[Serializable]` and even though this particular field isn't serialized, it feels like we are overloading that class as an object with both data and logic. It might actually be that it isn't being sent over the wire (I thought it was). In that case, we can remove `[Serializable]`.

Julian's argument seems to boil down to "this is implementation specific" / "this makes too many assumptions about implementations which we do not control". I think those are valid. Your counter points seems to be "implementations must conform to the spec, making those assumptions requirements". There are two aspects to this:
1. Is the `IServiceProvider` conforming? Yes, we only support conforming service providers.
2. Is the `IGrainActivator` conforming? Well... what does conforming mean here?

Users are expected to be able to inject their own implementation for `IGrainActivator`, but we do not have any way for them to know how to create one correctly. It almost seems that we need to create a test kit for our own spec for `IGrainActivator` to check for 'conforming grain activators'.

If the user provided the following container implementation then it's (surprisingly?) not conforming:
```C#
public class ServiceProviderGrainActivator : IGrainActivator
{
    public object Create(IGrainActivationContext context)
        => context.ActivationServices.GetService(context.GrainType);

    public void Release(IGrainActivationContext context, object grain)
        => (grain as IDisposable)?.Dispose();
}
```

Even though a conforming `IServiceProvider` must be in use, we're not constructing the grain correctly and we do not know which constructor is being used: the user might have registered `MyGrain` with the conforming container via a lambda and that lambda can construct `MyGrain` in whatever whacky way it wants.

The fact that a user must be using a conforming container does not mean that we know for sure how a type is constructed. Maybe that's fine, but it needs to be acknowledged.

This is the issue I'm more interested in resolving (or becoming comfortable with): the higher-level question of how grain construction is performed and what is acceptable for an `IGrainActivator`.